File: NoUpdateChecker.js
/*
 * NoUpdateChecker.js
 * Frida 腳本 - 阻止 iOS App 檢查更新
 * 
 * 使用方法：
 * frida -U -f com.target.app -l NoUpdateChecker.js
 * 
 * 或者附加到運行中的 App：
 * frida -U -n "TargetApp" -l NoUpdateChecker.js
 */

console.log("[*] NoUpdateChecker loaded");

// ============================================================
// 方法 1: Hook NSURLSession - 阻擋 App Store API
// ============================================================

var NSURLSession = ObjC.classes.NSURLSession;
var NSURLSession$sharedSession = NSURLSession.sharedSession;

if (NSURLSession) {
    Interceptor.replace(NSURLSession['- dataTaskWithRequest:completionHandler:'].implementation, 
        new NativeCallback(function(request, completionHandler) {
            var url = request.URL().absoluteString();
            
            // 阻擋 Apple App Store API
            if (url.indexOf("itunes.apple.com") !== -1 ||
                url.indexOf("apps.apple.com") !== -1 ||
                url.indexOf("lookup") !== -1 ||
                url.indexOf("softwareVersion") !== -1) {
                
                console.log("[NoUpdate] Blocked: " + url);
                
                // 返回空的 404 響應
                var NSData = ObjC.classes.NSData;
                var NSHTTPURLResponse = ObjC.classes.NSHTTPURLResponse;
                var response = NSHTTPURLResponse.alloc().initWithURL_statusCode_HTTPVersion_headerFields(
                    request.URL(), 404, "1.1", NSDictionary.alloc().init()
                );
                
                completionHandler(NSData.data(), response, NULL);
                return;
            }
            
            return originalDataTask(request, completionHandler);
        }, 'void', ['pointer', 'pointer'])
    );
}

// 保存原始方法
var originalDataTask = NSURLSession['- dataTaskWithRequest:completionHandler:'].implementation;

// ============================================================
// 方法 2: Hook NSBundle - 阻止更新信息獲取
// ============================================================

var NSBundle = ObjC.classes.NSBundle;

if (NSBundle) {
    var originalInfoDictionary = NSBundle['- infoDictionary'].implementation;
    
    Interceptor.replace(originalInfoDictionary, new NativeCallback(function(self, sel) {
        var result = originalInfoDictionary.call(self, sel);
        
        if (result) {
            // 移除更新相關的 key
            // 這裡只是示例，實際可能需要更精確
            console.log("[NoUpdate] infoDictionary accessed");
        }
        
        return result;
    }, 'pointer', ['pointer', 'pointer'])
    );
}

// ============================================================
// 方法 3: Hook UIApplication - 阻止打開 App Store
// ============================================================

var UIApplication = ObjC.classes.UIApplication;

if (UIApplication) {
    var originalOpenURL = UIApplication['- openURL:options:completionHandler:'];
    
    if (originalOpenURL) {
        Interceptor.replace(originalOpenURL.implementation, new NativeCallback(function(self, sel, url, options, completion) {
            var urlStr = url.toString();
            
            // 攔截 App Store 連結
            if (urlStr.indexOf("itunes.apple.com") !== -1 ||
                urlStr.indexOf("apps.apple.com") !== -1 ||
                urlStr.indexOf("itms-apps") !== -1) {
                
                console.log("[NoUpdate] Blocked URL: " + urlStr);
                
                if (completion) {
                    completion(false);
                }
                return;
            }
            
            return originalOpenURL.call(self, sel, url, options, completion);
        }, 'void', ['pointer', 'pointer', 'pointer', 'pointer', 'pointer'])
        );
    }
}

// ============================================================
// 方法 4: Hook SKStoreProductViewController
// ============================================================

var SKStoreProductViewController = ObjC.classes.SKStoreProductViewController;

if (SKStoreProductViewController) {
    var originalLoadProduct = SKStoreProductViewController['- loadProductWithParameters:completionBlock:'];
    
    if (originalLoadProduct) {
        Interceptor.replace(originalLoadProduct.implementation, new NativeCallback(function(self, sel, parameters, completionBlock) {
            console.log("[NoUpdate] Blocked SKStoreProductViewController");
            
            if (completionBlock) {
                var NSError = ObjC.classes.NSError;
                var error = NSError.errorWithDomain_code_userInfo(
                    ObjC.classes.NSString.alloc().initWithUTF8String("NoUpdateChecker"),
                    404,
                    NSDictionary.alloc().init()
                );
                completionBlock(error);
            }
        }, 'void', ['pointer', 'pointer', 'pointer', 'pointer'])
        );
    }
}

// ============================================================
// 方法 5: 阻止 App Store 服務 (NSUserDefaults)
// ============================================================

var NSUserDefaults = ObjC.classes.NSUserDefaults;

if (NSUserDefaults) {
    var originalObjectForKey = NSUserDefaults['- objectForKey:'];
    
    Interceptor.replace(originalObjectForKey.implementation, new NativeCallback(function(self, sel, key) {
        var keyStr = key.toString();
        
        // 阻止 Store 相關的 Key
        if (keyStr.indexOf("Store") !== -1 || 
            keyStr.indexOf("kSStore") !== -1 ||
            keyStr.indexOf("com.apple.AppStore") !== -1) {
            
            console.log("[NoUpdate] Blocked NSUserDefaults key: " + keyStr);
            return NULL;
        }
        
        return originalObjectForKey.call(self, sel, key);
    }, 'pointer', ['pointer', 'pointer'])
    );
}

console.log("[*] NoUpdateChecker active - Update checks will be blocked");

File: NoUpdateChecker.m
/*
 * NoUpdateChecker.dylib
 * 阻止 iOS App 檢查更新
 * 
 * 使用方法：
 * 1. 編譯成 .dylib
 * 2. 通過 Frida 注入或越獄環境加載
 */

#import <Foundation/Foundation.h>
#import <UIKit/UIKit.h>

// ============================================================
// 方法 1: Hook NSBundle - 阻止獲取更新資訊
// ============================================================

%hook NSBundle

- (NSDictionary *)infoDictionary {
    NSDictionary *original = %orig;
    
    // 移除 App Store 更新相關的鍵
    NSMutableDictionary *modified = [original mutableCopy];
    
    // 移除更新檢查相關的 key
    [modified removeObjectForKey:@"NSAppTransportSecurity"];
    [modified removeObjectForKey:@"UIAppFonts"];
    
    return modified;
}

%end

// ============================================================
// 方法 2: Hook UIApplication - 阻止更新彈窗
// ============================================================

%hook UIApplication

- (BOOL)openURL:(NSURL *)url options:(NSDictionary<UIApplicationOpenURLOptionsKey,id> *)options 
    completionHandler:(void (^)(BOOL))completion {
    
    // 攔截 App Store 連結
    if ([url.absoluteString containsString:@"itunes.apple.com"] ||
        [url.absoluteString containsString:@"apps.apple.com"] ||
        [url.absoluteString containsString:@"itms-apps"]) {
        
        NSLog(@"[NoUpdate] Blocked App Store URL: %@", url.absoluteString);
        
        if (completion) {
            completion(NO);
        }
        return YES; // 返回 YES 表示已處理，不進行實際打開
    }
    
    return %orig(url, options, completion);
}

%end

// ============================================================
// 方法 3: Hook NSUserDefaults - 阻止 App Store 檢查
// ============================================================

%hook NSUserDefaults

- (id)objectForKey:(NSString *)defaultName {
    // 阻止 Store 相關的 Key
    if ([defaultName containsString:@"Store"] || 
        [defaultName containsString:@"kSStore"] ||
        [defaultName containsString:@"com.apple.AppStore"]) {
        
        NSLog(@"[NoUpdate] Blocked NSUserDefaults key: %@", defaultName);
        return nil;
    }
    
    return %orig(defaultName);
}

%end

// ============================================================
// 方法 4: Hook NSURLSession - 阻擋 App Store API 請求
// ============================================================

%hook NSURLSession

- (NSURLSessionDataTask *)dataTaskWithRequest:(NSURLRequest *)request 
    completionHandler:(void (^)(NSData * _Nullable, NSURLResponse * _Nullable, NSError * _Nullable))completionHandler {
    
    NSString *url = request.URL.absoluteString;
    
    // 阻擋 Apple App Store API
    if ([url containsString:@"itunes.apple.com"] ||
        [url containsString:@"apps.apple.com"] ||
        [url containsString:@"Buy"] ||
        [url containsString:@"lookup"] ||
        [url containsString:@"softwareVersion"]) {
        
        NSLog(@"[NoUpdate] Blocked App Store API: %@", url);
        
        // 返回空響應
        NSData *emptyData = [NSData data];
        NSURLResponse *emptyResponse = [[NSHTTPURLResponse alloc] 
            initWithURL:request.URL 
            statusCode:404 
            HTTPVersion:@"1.1" 
            headerFields:@{}];
        
        completionHandler(emptyData, emptyResponse, nil);
        
        // 返回一個空任務
        return nil;
    }
    
    return %orig(request, completionHandler);
}

%end

// ============================================================
// 方法 5: Hook SKStoreProductViewController (如果有)
// ============================================================

%hook SKStoreProductViewController

- (void)loadProductWithParameters:(NSDictionary<NSString *,id> *)parameters 
    completionBlock:(void (^)(NSError * _Nullable))block {
    
    NSLog(@"[NoUpdate] Blocked SKStoreProductViewController load");
    
    // 返回錯誤阻止加載
    NSError *error = [NSError errorWithDomain:@"NoUpdateChecker" 
                                         code:404 
                                     userInfo:@{NSLocalizedDescriptionKey: @"Update blocked"}];
    if (block) {
        block(error);
    }
    return;
}

%end

// ============================================================
// 初始化
// ============================================================

%ctor {
    NSLog(@"[NoUpdateChecker] Loaded - Update checks disabled");
    NSLog(@"[NoUpdateChecker] This tweak blocks iOS app update checks");
}